occ commands for Impersonate (#646)
owncloud/docs-server: modules/admin_manual/pages/configuration/general_topics/impersonate_users.adoc

When you're ready to stop impersonating the user, log out and you will return to your normal user session.

> == Define Who can use Impersonate

Reviewer: ```suggestion
== Define Who can Use Impersonate
```

> == Define Who can Configure Impersonate

Reviewer 2: ```suggestion
== Define Who can use Impersonate
```

Actually this defines the users who have the Impersonate app "enabled" for them. (all users, or just the users in these groups)

For a user who is not an admin or sub-admin, they can be impersonated (but cannot impersonate anyone).

An admin can impersonate any of the users in the groups.

A sub-admin can impersonate users in their group(s) (if those extra settings are enabled).

Only admins can "configure Impersonate" (i.e. set the various system-wide settings for the Impersonate app)

Author: I do not get it...
Where is the difference in between:
`this defines the users who have the Impersonate app "enabled" for them`
`A sub-admin can impersonate users in their group(s)`
because in both settigs you can define groups (and only groups), but the first one is at the app section and the second one is defined at the user section at the 3 radiobuttons.
I have tested this and if you add a group in one part it does not show up at the other one

Reviewer 2: They are 2 completely separate things:
1) gives "access" to the app for whole groups of users (the potential to do impersonating or be impersonated)
then step 2) give rights to sub-admins to do impersonating

> * When you have installed and enabled the Impersonate App, you can define if and which groups are granted to _configure_ impersonation. To do so, go to menu:Your Username[Settings > Admin > Apps]. Members of the groups specified can configure impersonation, see xref:define-who-is-granted-to-impersonate[Define Who is Granted to Impersonate]

Reviewer 2: This all needs re-wording. The groups are the groups that "have access" to the Impersonate app. The users in those groups are the users who can do impersonation and can be impersonated. Which users can be an "impersonator" and which are potential "impersonatees" is based on the other settings.

Author: Agreed, as I see, not only this section needs rewording then, difference is still not clear to me, very complicated

Reviewer 2: Many apps have the ability to be enabled just for particular groups. In most cases the consequences are easy to imagine - example, enable the `comments` app for groups `group1` `group2` and `group3`. Users in any of those groups are shown existing comments and can make new comments, edit comments, delete comments. Other users on the system do not see any comments - for them it is just as if the comments app is disabled.

For the Impersonate app it is "the same thing" - but the consequences need to be thought about for a bit to understand. If the user doing the impersonating and the user being impersonated both have the app "enabled" then the impersonation attempt can go ahead, subject to any other settings/configs/rules of the Impersonate app.

If the Impersonate app is not enabled for the person who wants to do the impersonating, then they can't even get started using the app.

If the Impersonate app is not enabled for the person who would be impersonated, then the attempt at impersonation will fail.

But how to make that clear for readers?

Author: I am getting headaches :exploding_head:

Author: Lets do some examples to make it visible:
user admin is member of admin-g
user martin and phil are member of users-g and do not have any other membership
user phil is in addition group admin of the group users-g

**E1**
app enabled for all, no groups defined
radiobutton 3 (group admins for specific groups) + set group users-g
= phil can impersonate martin

**E2**
app enabled for users-g
radiobutton 3 (group admins for specific groups) + set group users-g
= phil can impersonate martin

**E3**
app enabled for admin-g
radiobutton 3 (group admins for specific groups) + set group users-g
= ??
the groups admin-g and user-g do not have any overlapping, therefore this would mean no impersonation possible

**E4**
app enabled for admin-g
radiobutton 2 (all group admins)
= ??
the groups admin-g and user phil do not have any group overlapping, therefore this would mean no impersonation possible

**E5**
app enabled for admin-g
radiobutton 1 (only owncloud admins)
= ??
the groups admin-g and any user do not have any group overlapping, therefore this would mean no impersonation possible

Compared to other apps, we have here the possibility of two group definition which may not have any overlapping. Because as long we stay in only one group, this is more clear. The question, is my view above correct?

Reviewer 2: E3 - correct - to be allowed to impersonate the app has to be "enabled" for the group(s) concerned, and then if the user is a sub-admin one of the "permissions for sub-admins" options needs to be selected.
E4 - correct - if the app is just enabled for the admin-g then admins can impersonate within that group, but to no-one else.
E5 - similar to E4 - admin-g are impersonating each other but no-one else.

The thing is to:
1) enable the app for all the groups that are allowed to be involved in Impersonation in any way (or enable for everyone - which is probably the normal use-case)

That allows real admins to do impersonating.

2) enable sub-admins to do Impersonating withing the group(s) that they manage - the radio-button settings.

That expands the Impersonation possibilities so that sub-admins can do Impersonating in a controlled way.

It makes no sense to enable (2) if (1) is severely restricted. First enable (1) as widely as wanted, then add on option (2) as needed.

When you're ready to stop impersonating the user, log out and you will return to your normal user session.

== Group Concept for Impersonation

> Impersonate has two layers defining groups for impersonation. It is important to understand, that both, the impersonator and the impersonee must be granted for the impersonation app. This means, if any groups are defined at the two layers, they must have an overlapping member area.

Reviewer: ```suggestion
Impersonate has two layers defining groups for impersonation. It is important to understand, that both, the impersonator and the user to be impersonated must have the impersonation app enabled. This means, if any groups are defined at the two layers, they must have an overlapping member area.
```

> . The first layer is defined at the level of the app, see xref:define-who-can-use-impersonate[Define Who can Use Impersonate]. This layer can be seen as a general area. If you do not want to have any restrictions, uncheck the checkmark which means all users are eligable.

Reviewer: ```suggestion
. The first layer is defined at the level of the app, see xref:define-who-can-use-impersonate[Define Who can Use Impersonate]. This layer can be seen as a general area. If you do not want to have any restrictions, uncheck the check box, which means all users are eligible.
```

> . The second layer is defined at the level where the one who can impersonate and the possible impersonees are configured, see xref:define-who-is-granted-to-impersonate[Define Who is Granted to Impersonate]. This layer can be seen as the detailed area.

Reviewer: ```suggestion
. The second layer is defined at the level where the one who can impersonate and who can be impersonated are configured. See xref:define-who-is-granted-to-impersonate[Define Who is Granted to Impersonate]. This layer can be regarded as the detailed area.
```

> As a rule of thumb, first enable (1) as widely as wanted, then add on option (2) as needed.

Reviewer: ```suggestion
As a rule of thumb, first enable (1) as widely as wanted, then add option (2) as needed.
```

> NOTE: If an impersonator is granted to impersonate but he cannot impersonate all or particular users, then the usual case for this situation is a lack of an overlapping area of groups defined.

Reviewer: ```suggestion
NOTE: If an impersonator is granted the right to impersonate but cannot impersonate all or particular users, then the reason is usually that the impersonator is not a member of the respective groups.
```

== Define Who can Use Impersonate

> * When you have installed and enabled the Impersonate App, you can define if and which groups are granted to _configure_ impersonation. To do so, go to menu:Your Username[Settings > Admin > Apps]. Members of the groups specified can configure impersonation, see xref:define-who-is-granted-to-impersonate[Define Who is Granted to Impersonate]

Reviewer: ```suggestion
* If you have installed and enabled the Impersonate App, you can define if any groups are granted the right to _configure_ impersonation. To do so, go to menu:Your Username[Settings > Admin > Apps]. Members of the groups specified can configure impersonation, see xref:define-who-is-granted-to-impersonate[Define Who is Granted to Impersonate]
```
